index.js

```diff
@@ -58,6 +84,139 @@ class DbFacility extends Base {
       }
     ], cb)
   }
+
+  /**
+   * @note Method does not support nested transactions
+   * @see https://www.npmjs.com/package/mysql#pooling-connections
+   * @see https://www.npmjs.com/package/mysql#transactions
+   *
+   * @param {(conn: { cli: mysql.PoolConnection }, txFuncCb: (err: Error|undefined) => void) => void} func
+   * @param {(err: DbTransactionError|undefined) => void} cb
+   */
+  runTransaction (func, cb) {
+    /** @type {mysql.PoolConnection} */
+    let conn = null
+    let txStarted = false
+    let txCommited = false
+    let txReverted = false
+
+    async.series({
+      conn: (next) => {
+        this.cli.getConnection((err, res) => {
+          if (err) return next(err)
+
+          conn = res // intentional
+          return next()
+        })
+      },
+      begin: (next) => {
+        conn.beginTransaction(next)
+      },
+      execute: (next) => {
+        txStarted = true
+        func({ cli: conn }, next)
+      },
+      commit: (next) => {
+        conn.commit(next)
+      },
+      release: (next) => {
+        txCommited = true
+        try {
+          conn.release()
+          return next()
+        } catch (err) {
+          return next(err)
+        }
+      }
+    }, (txFailureErr) => {
+      if (!txFailureErr) return cb()
+
+      if (txStarted && !txCommited) {
+        return async.series({
+          rollback: (next) => {
+            conn.rollback(next)
+          },
+          release: (next) => {
+            try {
+              txReverted = true
+              conn.release()
+              return next()
+            } catch (err) {
+              return next(err)
+            }
+          }
+        }, (rollbackErr) => {
+          if (rollbackErr) {
+            // force cleanup session
+            console.error(new Date().toISOString(), rollbackErr)
+
+            try {
+              conn.destroy()
+            } catch (destroyErr) {
+              console.error(new Date().toISOString(), destroyErr)
+            }
+          }
+
+          const cbErr = new DbTransactionError('ERR_TX_FLOW_FAILURE', txFailureErr, txStarted, txCommited, txReverted)
+          return cb(cbErr)
+        })
+      }
+
+      const cbErr = new DbTransactionError('ERR_TX_FLOW_FAILURE', txFailureErr, txStarted, txCommited, txReverted)
+      return cb(cbErr)
+    })
+  }
+
+  /**
+   * @note Method does not support nested transactions
+   * @see https://www.npmjs.com/package/mysql#pooling-connections
+   * @see https://www.npmjs.com/package/mysql#transactions
+   *
+   * @param {(conn: { queryAsync: Function, cli: mysql.PoolConnection }) => Promise<void>} func
+   */
+  async runTransactionAsync (func) {
```

Comment:
yep I initially thought of doing that but it would be difficult to maintain both approaches in same function, main problem is func param, if it contains cb then it would be difficult to handle both approaches

Reply:
this way although code is similar seems more manageable and more resistant to possible bugs